uniforce.lua

```diff
@@ -0,0 +1,259 @@
+  local assigned_items = {} -- assigned item ids mapped to their armor location
+
+  for loc, specs in pairs( squad_position.uniform ) do -- indexed by armor location
+    for i, spec in pairs(specs) do
+      for i, assigned in pairs( spec.assigned ) do
+        -- Include weapon and shield so we don't drop them later
+        assigned_items[ assigned ] = loc
+      end
+    end
+  end
+
+  -- Figure out which assigned items are currently not being worn
+
+  local present_ids = {} -- map of item ID to item object
+  local missing_ids = {} -- map of item ID to armor location
+  local missing_locs = {} -- map of armor locations to true/nil
+  for u_id, loc in pairs(assigned_items) do
+    if worn_items[ u_id ] == nil then
+      local item = find_item_by_ID( u_id )
+      dfhack.println("Unit "..unit_name.." is missing an assigned "..loc.." item, object #"..u_id.." '"..utils.getItemDescription(item).."'" )
+      missing_ids[ u_id ] = loc
+      missing_locs[ loc ] = true
+      if args.free then
+        to_drop[ u_id ] = item
+      end
+    else
+      present_ids[ u_id ] = worn_items[ u_id ]
+    end
+  end
+
+  -- Figure out which worn items should be dropped
+
+  -- First, figure out which body parts are covered by the uniform pieces we have.
+  local covered = {} -- map of body part id to true/nil
+  for id, item in pairs( present_ids ) do
+    local loc = assigned_items[ id ]
+    if loc ~= 'shield' and loc ~= "weapon" then -- shields and weapons don't "cover" the bodypart they're assigned to.
+     covered[ worn_parts[ id ] ] = id
+    end
+  end
+
+  if multi then
+    covered = {} -- Don't consider current covers - drop for anything which is missing
+  end
+
+  -- Figure out body parts which should be covered but aren'y
+  local uncovered = {}
+  for part, loc in pairs( PART_TO_POSITION ) do
+    if not covered[ part ] then
+      -- Only mark it "uncovered" if we're nominally supposed to have something there.
+      if missing_locs[ loc ] then
+        uncovered[ part ] = true
+      end
+    end
+  end
+
+  -- Drop everything (except uniform pieces) from body parts which should be covered but aren't
+  for w_id, item in pairs(worn_items) do
+    if assigned_items[ w_id ] == nil then -- don't drop uniform pieces (including shields, weapons for hands)
+      if uncovered[ worn_parts[ w_id ] ] then
+        dfhack.println("Unit "..unit_name.." potentially has object #"..w_id.." '"..utils.getItemDescription(item).."' blocking their uniform "..PART_TO_POSITION[ worn_parts[ w_id ] ])
+        if args.drop then
+          to_drop[ w_id ] = item
+        end
+      end
+    end
+  end
+
+  return to_drop
+end
+
+
+function do_drop( item_list )
+  if item_list == nil then
+    return nil
+  end
+
+  for id, item in pairs(item_list) do
+    local pos = get_item_pos(item)
+    if pos == nil then
+      dfhack.println("Could not find drop location for item #"..id.."  "..utils.getItemDescription(item))
```

Comment:
dfhack.println is aliased to 'print', which is shorter and cleaner

Reply:
`dfhack.printerr` could also be an alternative for messages that are errors (it does not have an alias, though). I'm guessing `qerror` would not be appropriate since you don't want to actually raise an error here (since you're processing a list).